models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,486 @@
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  double_t h = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -h / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -h / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, h );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, h );
+
+
+  V_.P33_ = std::exp( -h / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.sfa_stc_initialized_ )
+  {
+    V_.P_sfa_.clear();
+    V_.P_stc_.clear();
+    S_.sfa_elems_.clear();
+    S_.stc_elems_.clear();
+
+    for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.P_sfa_.push_back( std::exp( -h / P_.tau_sfa_[ i ] ) );
+      S_.sfa_elems_.push_back( 0.0 );
+    }
+
+    for ( size_t i = 0; i < P_.tau_stc_.size(); i++ )
+    {
+      V_.P_stc_.push_back( std::exp( -h / P_.tau_stc_[ i ] ) );
+      S_.stc_elems_.push_back( 0.0 );
+    }
+
+    S_.sfa_stc_initialized_ = true;
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_psc_exp::update( Time const& origin,
+  const long_t from,
+  const long_t to )
+{
+
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  double_t q_temp_;
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    // exponential decaying stc and sfa elements
+    q_temp_ = 0;
+    for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.stc_elems_[ i ];
+
+      S_.stc_elems_[ i ] = V_.P_stc_[ i ] * S_.stc_elems_[ i ];
+    }
+
+    S_.stc_ = q_temp_;
+
+    q_temp_ = 0;
+    for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.sfa_elems_[ i ];
+
+      S_.sfa_elems_[ i ] = V_.P_sfa_[ i ] * S_.sfa_elems_[ i ];
+    }
+
+    S_.sfa_ = q_temp_ + P_.V_T_star_;
+
+    // exponential decaying PSCs
+    S_.i_syn_ex_ *= V_.P11ex_;
+    S_.i_syn_in_ *= V_.P11in_;
+
+    S_.i_syn_ex_ += B_.spikes_ex_.get_value( lag );
+    S_.i_syn_in_ += B_.spikes_in_.get_value( lag );
+
+    if ( S_.r_ref_ == 0 ) // neuron not refractory, so evolve V
+    {
+
+      if ( S_.add_stc_sfa_ )
```

Comment:
By adding jumps to the STC and SFA here, you delay the onset of spike-triggered currents and threshold adaptation until the end of the refractory period. My understanding of the paper is that both are triggered instantaneously when the spike occurs. Thus, the steps should occur in conjunction with spike emission, i.e., in the place where you now set `add_stc_sfa_` to `true`.  This member could then disappear.